common/workunit/workunit.cpp

```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    const unsigned numStats = aggregateKindsMapping.numStatistics();
+    for (unsigned i=0; i<numStats; ++i)
+        filter.addOutputStatistic(aggregateKindsMapping.getKind(i));
+    filter.setDepth(1,3); // 1=global, 2=workflow, 3=graph
+    filter.setSources(SSFsearchGlobalStats);
+    filter.setIncludeNesting(0);
+    filter.finishedFilter();
+
+    StatsCollectionAggregatesLoader aggregatesLoader(statsCollection);
+    Owned<IConstWUScopeIterator> iter = &workunit.getScopeIterator(filter);
+    ForEach(*iter)
+        iter->playProperties(aggregatesLoader);
+}
+
+// getCollectionForUpdate() returns IStatisticCollection for the given subgraph
+// if clearStats==true then the existing stats are cleared for the given scope and descendants
+IStatisticCollection * GlobalStatisticCollection::getCollectionForUpdate(StatisticCreatorType creatorType, const char * creator, unsigned wfid, const char *graphName, unsigned sgId, bool clearStats)
+{
+    StatsScopeId graphScopeId;
+    verifyex(graphScopeId.setScopeText(graphName));
+    StatsScopeId wfScopeId(SSTworkflow, wfid);
+    StatsScopeId sgScopeId(SSTsubgraph, sgId);
+
+    bool wasCreated;
+    IStatisticCollection * sgScopeCollection = statsCollection->ensureSubScopePath({wfScopeId,graphScopeId, sgScopeId}, wasCreated);
+    if (clearStats && !wasCreated)
+        sgScopeCollection->clearStats();
+    // Marking the collection dirty here is not ideal.  It would be better to have a call to IStatisticCollection::setStatistic mark the scope as dirty.
+    // However, this would be inefficient as each call to IStatisticCollection::setStatistic would require the dirty flag to be set for all parent scopes.
+    sgScopeCollection->markDirty();
+    return createRootStatisticCollection(creatorType, creator, wfScopeId, graphScopeId, sgScopeCollection);
+}
+
+// Recalculate aggregates for global, workflow and graph scopes
+bool GlobalStatisticCollection::refreshAggregates()
+{
+    return statsCollection->refreshAggregates(aggregateKindsMapping);
+}
+
+// Recalculate aggregates and then write the aggregates to global stats (dali)
+void GlobalStatisticCollection::updateAggregates(IWorkUnit *wu)
+{
+    class StatisticsAggregatesWriter : implements IStatisticVisitor
+    {
+        const StatisticsMapping & aggregateKindsMapping;
+        const unsigned numStats;
+        Linked<IWorkUnit> wu;
+    public:
+        StatisticsAggregatesWriter(IWorkUnit * _wu, const StatisticsMapping & _aggregateKindsMapping): wu(_wu), aggregateKindsMapping(_aggregateKindsMapping), numStats(aggregateKindsMapping.numStatistics()) {}
+
+        virtual bool visitScope(const IStatisticCollection & cur)
         {
-            it->next();
+            switch (cur.queryScopeType())
+            {
+            case SSTglobal:
+            case SSTworkflow:
+            case SSTgraph:
+                for (unsigned i=0; i<numStats; ++i)
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
```

Comment:
Related to refreshAggregates / isDirty comments, and since regressAggretates is always done in order to push updates, should both these recursive descents into scopes to check for 'dirty' be combined?
i.e. so that updateAggregates only recalculates aggregates it needs to (if the stat is dirty) and writes them at that point (but commits to Dali at end).

Reply:
I've combined the refreshAggregates with updateAggregates so that now refreshAggregate can use a callback function to update an modified aggregate.